### ISSUE TITLE
Document Git workflow (was "Zender needs hand-holding with Git)

### ISSUE THREAD

**Issue by czender** [terraref/computing-pipeline#233]
### Description

I need GitHub help before I can update hyperspectral workflow.
With the old permissions, I would edit locally the master branch of
computing-pipeline, and then push that directly to GitHub.
With the new permissions, GitHub refuses to let me push master, and
instead requires that I submit a pull request.

### Details

I do not know how to do this, and I need help from a Git-guru to
reformulate my workflow. Would someone please call and walk me through
the recommended Git procedures? I will write it down and make sure it works
for me and use it in the future. Again, I have changes locally on
master that I cannot push upstream. Please tell me _exactly_ what the recommended 
Git workflow is, and help me follow it the first time until it succeeds.

### Completion Criteria   

* [ ] My current local modifications to master successfully posted as PR
* [ ] Reciped written down to follow next time
* [ ] Next batch of mods successfully upstreamed according to recipe, without handholding
* [ ] CLI and GUI Git Workflow updated / added to documentation

**Comment by dlebauer** [comment]
@czender 

Assuming you have already committed your changes

1. create new branch
```sh
git checkout -b newfeaturebranch ## create a new branch
git push origin newfeature branch
```

2. go to github, open the branches tab, and under 'active branches' create pull request

**Comment by czender** [comment]
Thanks both, I am trying David's recipe for now...

**Comment by czender** [comment]
apparently my first PR (for branch swir_273_channels) was approved but not merged yet and now @dlebauer wants another change so then i further edited branch swir_273_channels and commited new stuff to determine sensor type from _metadata.json file instead of wvl_nbr and created new branch called  sensor_type_from_metadata for that, and made PR for that. Is it a good idea to issue this command locally:
git branch --set-upstream-to=origin/master swir_273_channels
git branch --set-upstream-to=origin/master sensor_type_from_metadata
so that if/when the PRs are merged these branches will have the latest and greatest?
Or would those commands do something unexpected?

**Comment by max-zilla** [comment]
I haven't used those commands, but typically once this is merged into master you don't need these branches anymore. You can switch back to master (latest and greatest) and branch off of that for future work.

I think I confused things just now by approving your 2nd pull request (sensor type from metadata) and merging it without realizing it would also merge the swir_273 branch. In general I think it would be easiest for you if, in the situation you described, you just add the sensor_type_from_metadata code directly to the swir_273 branch and not make another branch/pull request. The original PR will automatically update with the new stuff and we can re-comment, re-review the changes, etc, and continue in this fashion until we're happy.

I will be a bit slower on the merge for these PRs from now on, I am not used to others being on the review list as well :)

**Comment by czender** [comment]
i created the second branch because i didn't think that any subsequent (to my PR) commits to the first branch would be applied. because they did not exist at the time of the PR, and were not described in the original PR. but the new mods were written in response to a review of the PR, so maybe I should have been more optimistic. 

please continue to be fast at merging because i need feedback in real time today in order to learn this new github workflow. that's what i mean by handholding. :) i have the time to learn this today but not next week.

**Comment by max-zilla** [comment]
Luckily subsequent commits DO retroactively apply to pull requests, and I believe it'll even mark outdated comments in the PR if you're looking at a comment for a line of code that's been updated later.

**Comment by dlebauer** [comment]
and in general, you could develop on a 'feature branch' indefinitely, only merging to master when it is ready to deploy. https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

We need to document a common workflow for the team, but it will follow that one.

**Comment by czender** [comment]
according to this feature branch, @dlebauer, i should push my newfeature branch upstream with
git push -u origin newfeature
instead of 
git push origin newfeature
as you recommend above. Is my understanding correct?
I'm not sure what the operational difference is between these two styles. Any insight to offer?

**Comment by dlebauer** [comment]
-u is alias for--set-upstream It sets the upstream branch so you can `git pull` from it without specifying where from. It's a shortcut. I prefer to be explicit 

see

http://explainshell.com/explain?cmd=git+push+-u+origin+newfeature

**Comment by czender** [comment]
I have a workable recipe now. Will re-open later if necessary. Thanks for the help Max and David.
